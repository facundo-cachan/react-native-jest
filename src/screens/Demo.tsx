/**
 * Demo Screen
 * Use like template to create news screens
 *
 * @screen
 */

import { Text } from 'react-native'

const DemoScreen = () => (
  <Text>Demo Screen</Text>
);

export default DemoScreen;
