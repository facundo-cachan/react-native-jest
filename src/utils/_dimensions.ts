import { Dimensions, PixelRatio } from 'react-native'

let { height, width } = Dimensions.get('screen')

/**
 * Converts provided width percentage to independent pixel (dp).
 * @param  {string} widthPercent The percentage of screen's width that UI element should cover
 *                               along with the percentage symbol (%).
 * @return {number}              The calculated dp depending on current device's screen width.
 */
const widthPercentageToDP = (widthPercent: string | number) => {
  // Parse string percentage input and convert it to number.
  const elemWidth =
    typeof widthPercent === 'number' ? widthPercent : parseFloat(widthPercent)
  // Use PixelRatio.roundToNearestPixel method in order to round the layout
  // size (dp) to the nearest one that correspons to an integer number of pixels.
  return PixelRatio.roundToNearestPixel((width * elemWidth) / 100)
}

/**
 * Converts provided height percentage to independent pixel (dp).
 * @param  {string} heightPercent The percentage of screen's height that UI element should cover
 *                                along with the percentage symbol (%).
 * @return {number}               The calculated dp depending on current device's screen height.
 */
const heightPercentageToDP = (heightPercent: string | number) => {
  // Parse string percentage input and convert it to number.
  const elemHeight =
    typeof heightPercent === 'number'
      ? heightPercent
      : parseFloat(heightPercent)

  // Use PixelRatio.roundToNearestPixel method in order to round the layout
  // size (dp) to the nearest one that correspons to an integer number of pixels.
  return PixelRatio.roundToNearestPixel((height * elemHeight) / 100)
}

/*
* Implement that piece of code to have orientation detection

Dimensions.get('screen').width < Dimensions.get('screen').height ? 'portrait' : 'landscape'
*/

export { height, width, widthPercentageToDP as wp, heightPercentageToDP as hp }
