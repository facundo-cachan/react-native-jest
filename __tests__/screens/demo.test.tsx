/* eslint-disable @typescript-eslint/ban-ts-comment */
import { afterEach, beforeEach, describe, expect, test } from '@jest/globals';
import { cleanup, screen, waitFor } from '@testing-library/react-native';

import Demo from '@screens/demo';
import { App, render, toJSON } from '../utils/customRender';

const props = {
  component: Demo,
  name: 'Demo'
};
describe('Demo screen', () => {
  afterEach(() => {
    cleanup();

    // Clean up the mocked fetch after each test
    // @ts-ignore
    global.fetch.mockClear();
  });
  beforeEach(() => {
    // Mock fetch with a resolved Promise that returns sample data
    // @ts-ignore
    jest.clearAllMocks();

    // @ts-ignore
    global.fetch = jest.fn().mockResolvedValue({
      json: () => Promise.resolve()
    });
    render(props);
  });
  describe('Render', () => {
    test('SnapShot Screen', () => {
      toJSON(<App {...props} />);
      expect(screen).toMatchSnapshot();
    });
    test('Render components & values on the screen', async () => {
      const goTo = screen.getByText(/Go to/);
      const byText = screen.getByText(/Screen/);
      const byPlaceholderText = screen.getByPlaceholderText(/Name/);
      waitFor(() => {
        expect(goTo).toBeTruthy();
        expect(byText).toBeTruthy();
        expect(byPlaceholderText).toBeTruthy();
      });
    });
  });
});
