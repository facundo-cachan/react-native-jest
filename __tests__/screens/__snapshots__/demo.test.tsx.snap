// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demo screen Render SnapShot Screen 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          pointerEvents="box-none"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={3}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          {
                            "alignItems": "center",
                            "backgroundColor": "#3cc6",
                            "justifyContent": "center",
                            "marginTop": 100,
                            "width": 750,
                          }
                        }
                      >
                        <Text
                          style={
                            [
                              {
                                "fontFamily": "Cochin",
                              },
                              undefined,
                            ]
                          }
                        >
                          Demo Screen
                        </Text>
                        <TextInput
                          accessibilityLabel="User Name"
                          keyboardType="default"
                          onChangeText={[Function]}
                          placeholder="User Name"
                          style={
                            [
                              {
                                "fontFamily": "Cochin",
                              },
                              undefined,
                            ]
                          }
                          value=""
                        />
                        <View
                          accessibilityLabel="Go to Details"
                          accessibilityRole="button"
                          accessibilityState={
                            {
                              "busy": undefined,
                              "checked": undefined,
                              "disabled": undefined,
                              "expanded": undefined,
                              "selected": undefined,
                            }
                          }
                          accessibilityValue={
                            {
                              "max": undefined,
                              "min": undefined,
                              "now": undefined,
                              "text": undefined,
                            }
                          }
                          accessible={true}
                          collapsable={false}
                          focusable={true}
                          onClick={[Function]}
                          onResponderGrant={[Function]}
                          onResponderMove={[Function]}
                          onResponderRelease={[Function]}
                          onResponderTerminate={[Function]}
                          onResponderTerminationRequest={[Function]}
                          onStartShouldSetResponder={[Function]}
                          style={
                            {
                              "opacity": 1,
                            }
                          }
                        >
                          <View
                            style={
                              [
                                {},
                              ]
                            }
                          >
                            <Text
                              style={
                                [
                                  {
                                    "color": "#007AFF",
                                    "fontSize": 18,
                                    "margin": 8,
                                    "textAlign": "center",
                                  },
                                ]
                              }
                            >
                              Go to Details
                            </Text>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
